### PR TITLE
Fix typing on `Tensor` inplace ops

### LINKED ISSUE #103375: Inplace binary ops on tensor subclasses can cause mypy error
### 🐛 Describe the bug

Using inplace binary ops on a subclass of `Tensor` will cause mypy error (e.g. `*=` shown below, same for other ops `+=` etc.)
```python
import torch
a = torch.nn.Parameter(torch.Tensor())
a *= 2
```
run `mypy`:
```
a.py:3: error: Result type of * incompatible in assignment
```
This is because of pyi definition `def __imul__(self, other: Any) -> Tensor: ...`, which means it always returns `Tensor` but not the subclass.

---

However, `**=` will not cause any mypy error.
```python
import torch
a = torch.nn.Parameter(torch.Tensor())
a **= 2
reveal_type(a)
```
but pyright (I use it through pylance) will give:
```
Type of "a" is "Any | Unknown | _NotImplementedType"
```
This is because `__ipow__` is not defined in the pyi, although `print(torch.Tensor.__ipow__)` shows a valid result.

### Versions

master

### PR DESCRIPTION
**Pull request by lkct**
Fixes #103375

Changes:
- Make `__iop__` return `Self` (the current mypy version does not support PEP 673 `Self`, so it's implemented by a `TypeVar`)
- Make `op_` return `Self` (e.g. `add_`)
- Sort `__iop__` and `__rop__` with cleanup
  - Sort them to have the same order as `__op__` (looks more clear)
  - Some ops actually not in `dir(torch._C._TensorBase)` and therefore removed (defined in `Tensor` instead)

### Example
```python
import torch
a = torch.nn.Parameter()
a *= 2
reveal_type(a.mul_(2))
print(a.mul_(2).__class__)  # torch.nn.parameter.Parameter
```
Previously:
```
a.py:3:1: error: Result type of * incompatible in assignment  [misc]
a.py:4:13: note: Revealed type is "torch._tensor.Tensor"
```
Now:
```
a.py:4:13: note: Revealed type is "torch.nn.parameter.Parameter"
```

### CODE REVIEW

File: tools/pyi/gen_pyi.py [hunk 1]
```diff
@@ -1023,6 +1021,19 @@ def gen_pyi(
...
     tensor_method_hints = []
     for name, hints in sorted(unsorted_tensor_method_hints.items()):
+        if (
+            name.startswith("__")
+            and name.endswith("__")
+            and name.strip("_") in inplace_binary_ops
+        ) or (
+            not name.startswith("_") and not name.endswith("__") and name.endswith("_")
+        ):
+            # _TensorBase.__iadd__() etc and _TensorBase.idd_() etc should return self
```
Comment (soulitzer) [inline tools/pyi/gen_pyi.py:1031]: typo: idd_ -> add_

**Review by soulitzer — COMMENTED** [review]
Thanks for the fix, just some small comments

File: tools/pyi/gen_pyi.py (hunk 1 above)
Comment (lkct) [reply tools/pyi/gen_pyi.py]: fixed

File: tools/pyi/gen_pyi.py [hunk 2]
```diff
@@ -146,37 +146,35 @@ def should_bind_method(python_func: PythonSignatureNativeFunctionPair) -> bool:
...
     "mod",
     "truediv",
     "matmul",
     "floordiv",
     "radd",
-    "rsub",
-    "rmul",
-    "rtruediv",
-    "rfloordiv",
-    "rpow",  # reverse arithmetic
```
[the remark this answers is not in the document]
Comment (lkct) [reply tools/pyi/gen_pyi.py:161]: I mentioned this in the last item in the description. Typing for these methods is another story, addressed in #103394.

To give some further information:
```python
>>> import torch  # nightly 20230611
>>> [a for a in dir(torch._C._TensorBase) if a.startswith('__r')]
['__radd__', '__rand__', '__reduce__', '__reduce_ex__', '__repr__', '__rmul__', '__ror__', '__rshift__', '__rxor__']
```

**Review by malfet — CHANGES_REQUESTED** [review]
Can you please add a test (say to `test_typing.py`) otherwise it can easily regress again.

Also, in PR description would be nice to add how mypy inferred type looked before and after the change

**Comment by lkct** [comment]
> Can you please add a test (say to `test_typing.py`) otherwise it can easily regress again.

Can you please give some hint on what this test should look like? It should be fed to mypy for static checking but not for runtime testing.

> Also, in PR description would be nice to add how mypy inferred type looked before and after the change

Will add soon.

**Comment by lkct** [comment]
> Can you please add a test (say to `test_typing.py`) otherwise it can easily regress again.

I found some examples.

> Also, in PR description would be nice to add how mypy inferred type looked before and after the change

Added

**Comment by lkct** [comment]
@malfet Help please!

I cannot get `test_typing.py` working locally. Can you please confirm this test actually works and is active?

Following is why I think this file does not work. Please point out if I'm wrong.

---

Firstly in this test file, it loads some reveal tests from this location.
https://github.com/pytorch/pytorch/blob/31ee1512d3108a2856f799250fcefc667c13832c/test/test_typing.py#L23-L24

And here is a test in the referred folder.
https://github.com/pytorch/pytorch/blob/31ee1512d3108a2856f799250fcefc667c13832c/test/typing/reveal/namedtuple.py#L12-L15

The function `qr` is defined here.
https://github.com/pytorch/pytorch/blob/31ee1512d3108a2856f799250fcefc667c13832c/torch/linalg/__init__.py#L2661

Where the `_linalg` is imported in this line of the same file.
https://github.com/pytorch/pytorch/blob/31ee1512d3108a2856f799250fcefc667c13832c/torch/linalg/__init__.py#L5

However, I cannot find `_linalg` from `torch._C`.

**Comment by lkct** [comment]
@malfet Can you please kindly provide some hints on what to do with the tests? Many thanks!

**Comment by huydhn** [comment]
For the context, `test_typing` is in the test blocked list atm https://github.com/pytorch/pytorch/blob/main/test/run_test.py#L187, so it's not even run on CI

File: tools/pyi/gen_pyi.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lkct) [reply tools/pyi/gen_pyi.py:161]: @soulitzer May I consider this resolved?

**Pull request closed without merging**